src/EventStore.Core/Services/RequestManager/LogNotificationService.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using EventStore.Common.Utils;
+
+namespace EventStore.Core.Services.RequestManager {
+	public class LogNotificationService : IDisposable {
+		//todo: allocate in a pool and reuse
+		public struct PositionNode {
+			public readonly long Position;
+			public readonly TaskCompletionSource WaitTask;
+			public PositionNode(long position) {
+				Position = position;
+				WaitTask = new TaskCompletionSource();
+			}
+		}
+		private readonly LinkedList<PositionNode> _registeredActions = new LinkedList<PositionNode>();
+		private long _logPosition;
+		private long _nextLogPosition;
+		private readonly Thread _thread;
+		private object _registerLock = new object();
+		private readonly AutoResetEvent _positionUpdated = new AutoResetEvent(true);
+		private readonly ManualResetEventSlim _stopped = new ManualResetEventSlim(false);
+		private readonly string _name;
+		private CancellationTokenSource _cancelSource;
+		private CancellationToken _canceled;
+		private bool _disposedValue;
+
+		public LogNotificationService(string name) {
+			_cancelSource = new CancellationTokenSource();
+			_canceled = _cancelSource.Token;
+			_thread = new Thread(Notify) { IsBackground = true, Name = name };
+			_thread.Start();
+			_name = name;
+		}
+		public void UpdateLogPosition(long position) {
+			if (Interlocked.Read(ref _nextLogPosition) <= position) {
+				Interlocked.Exchange(ref _nextLogPosition, position);
+				_positionUpdated.Set();
+			}
+		}
+		private void Notify() {
+			while (!_canceled.IsCancellationRequested) {
```

Comment:
I'm not familiar with what this code should do*, but it seems this is quite heavy on sync-primitives.
Can something like [Channels](https://devblogs.microsoft.com/dotnet/an-introduction-to-system-threading-channels/) be used to do the same thing but leass heavy?

<sub>\* just started to dig into the code base a little bit</sub>